python/tank/util/shotgun.py

```diff
@@ -312,25 +285,80 @@ def create_sg_connection(user="default"):
     :param user: Optional shotgun config user to use when connecting to shotgun, as defined in shotgun.yml
     :returns: SG API instance
     """
-    api_handle, _ = __create_sg_connection(__get_sg_config(), evaluate_script_user=False, user=user)
+    # get connection parameters
+    config_path = __get_sg_config()
+    config_data = __get_sg_config_data(config_path, user)
+    api_handle = __create_sg_connection(config_data)
     return api_handle
 
+g_app_store_connection = None
 def create_sg_app_store_connection():
     """
-    Creates a shotgun connection to the tank app store.
+    Creates a shotgun connection that can be used to access the toolkit app store.
 
-    returns a tuple: (api_handle, script_user_entity)
-
-    The second part of the tuple represents the
-    user that was used to connect to the app store,
-    as a standard sg entity dictionary.
+    :returns: (sg, dict) where the first item is the shotgun api instance and the second 
+              is an sg entity dictionary (keys type/id) corresponding to to the user used
+              to connect to the app store.
     """
     global g_app_store_connection
     
-    if g_app_store_connection is None:
-        g_app_store_connection = __create_sg_connection(__get_app_store_config(), evaluate_script_user=True)
+    # maintain a cache for performance
+    if g_app_store_connection:
```

Comment:
I feel like we should implement a simple memoize decorator. I've seen this caching pattern in quite a few locations and it would make the code cleaner if we had one.